### ISSUE TITLE
Why detach prototypes in "prototype_all" branch?

### ISSUE THREAD

**Issue by HDUyiming** [imatif17/Prototype-Mean-Teacher#14]
```
            loss_pro1 = F.cosine_similarity(self.prototype_s1, self.prototype_t).abs().mean().to(self.device)
            loss_pro2 = F.cosine_similarity(self.prototype_s2, self.prototype_t).abs().mean().to(self.device)
            loss_pro3 = F.cosine_similarity(self.prototype_s1, self.prototype_s2).abs().mean().to(self.device)
            loss = -(loss_pro1 + loss_pro2 + loss_pro3)/3
            self.prototype_t.detach_()
            self.prototype_s1.detach_()
            self.prototype_s2.detach_()

            loss_c = 0
            if self.contra:
                for i, j in itertools.combinations(list(range(len(self.prototype_s1))),2):
                    loss_c += F.cosine_similarity(self.prototype_s1[i].view(-1,1), self.prototype_s1[j].view(-1,1)).abs().mean().to(self.device)
                loss_c = loss_c/math.comb(len(self.prototype_s1), 2)
                self.prototype_s1.detach_()

            #print(f"Pro loss  :   {loss}")
            return loss, loss_c
```
Doesn't the computational graph of the model break after detach? In this way, the loss calculated from the prototype cannot be used to update the model through gradients.

**Comment by imatif17** [comment]
Hi! Now, looking into it, I think you're correct. Detach should come after loss_c loss, else the graph will be broken. I'll update the code, soon. Thanks!

**Comment by HDUyiming** [comment]
> Hi! Now, looking into it, I think you're correct. Detach should come after loss_c loss, else the graph will be broken. I'll update the code, soon. Thanks!

Thank you for your reply! I'd also like to ask what the impact would be if I don't use detach here. Why do we need to detach the prototype features?

**Comment by imatif17** [comment]
Prototypes are part of the model's parameter. If you don't use detach, they will receive gradient and will be updated with gradient descent. We don't want that, we want to compute prototypes using ROI features (not with gradient). detach cuts the prototypes from the computational graph, so they will not be updated using gradients. If you don't want to use detach, you can remove prototypes from the optimizer, or make their learning rate 0.

**Comment by imatif17** [comment]
Earlier I used to think that it was okay to use detach, but nowadays I think we should generally avoid detach. So, I'll try to remove detach from the code base and just remove prototypes from the optimizer.

**Comment by HDUyiming** [comment]
> Earlier I used to think that it was okay to use detach, but nowadays I think we should generally avoid detach. So, I'll try to remove detach from the code base and just remove prototypes from the optimizer.

Thank you for your enthusiastic response. I still want to consult you about the error reporting problem in DDP training. I've also read other issues and your new work (ACIA), but I still don't know how to modify this prototype calculation to avoid the error reporting problem in multi-card training.

Can you tell me in detail how to avoid DDP errors? It would be even better if you could provide code support.

**Comment by imatif17** [comment]
I will try to update the code to work with DDP soon.

**Comment by HDUyiming** [comment]
> I will try to update the code to work with DDP soon.

Thank you!
